src/sknnr/transformers/_cca.py

```diff
@@ -13,228 +22,85 @@ def zero_sum_columns(arr: NDArray) -> NDArray:
-@dataclass
-class InitCA:
-    Y_: NDArray
-    method: str = "CA"
-
-    @property
-    def tot(self):
-        return np.sum(self.Y_)
-
-    def _normalized(self):
-        return self.Y_ / self.tot
-
-    @property
-    def rw(self):
-        return np.sum(self._normalized(), axis=1)
-
-    @property
-    def cw(self):
-        return np.sum(self._normalized(), axis=0)
-
-    @property
-    def Y(self):
-        rc = np.outer(self.rw, self.cw)
-        return (self._normalized() - rc) / np.sqrt(rc)
-
-
-# Corresponds to vegan function "ordHead" in ordConstrained.R
-class OrdHead:
-    def __init__(self, transform):
-        self.transform = transform
-
-    @property
-    def tot_chi(self):
-        Y = self.transform.Y
-        return np.sum(Y * Y)
-
-    @property
-    def Ybar(self):
-        return self.transform.Y
-
-    @property
-    def grand_total(self):
-        return self.transform.tot
-
-    @property
-    def rowsum(self):
-        return self.transform.rw
-
-    @property
-    def colsum(self):
-        return self.transform.cw
-
-
-class ConstrainedOrdination:
-    Y_TRANSFORM = {
-        "CCA": InitCA,
-    }
+class ConstrainedOrdination(ABC):
     ZERO: float = math.sqrt(2.220446e-16)
 
-    @property
-    def transform(self):
-        return self.Y_TRANSFORM[self.method](self.Y)
-
-    @property
-    def head(self):
-        return OrdHead(self.transform)
-
-    @property
-    def dist_based(self) -> bool:
-        return self.transform.method == "DISTBASED"
-
-    @property
-    def rw(self):
-        return self.transform.rw
-
-    @property
-    def cw(self):
-        return self.transform.cw
-
-    @property
-    def env_center(self):
-        return np.average(self.X, axis=0, weights=self.rw)
-
-    @property
-    def X_scale(self):
-        X = self.X - self.env_center
-        return X * np.sqrt(self.rw)[:, np.newaxis]
-
-    @property
-    def qr(self):
-        return np.linalg.qr(self.X_scale)
-
-    @property
-    def Q(self):
-        return self.qr[0]
-
-    @property
-    def R(self):
-        return self.qr[1]
-
-    @property
-    def _least_squares(self):
-        Q, R = self.qr
-        right = Q.T @ self.transform.Y
-        return np.linalg.lstsq(R, right, rcond=None)
-
-    @property
-    def Y_fit(self):
-        return self.X_scale @ self._least_squares[0]
-
-    @property
-    def sol(self):
-        return np.linalg.svd(self.Y_fit, full_matrices=False)
-
-    @property
-    def rank(self):
-        ls_rank = self._least_squares[2]
-        pos_s = np.sum([self.sol[1] > self.ZERO])
-        return min(ls_rank, pos_s)
-
-    @property
-    def _U(self):
-        return self.sol[0][:, : self.rank]
-
-    @property
-    def S(self):
-        return np.square(self.sol[1])[: self.rank]
-
-    @property
-    def _V(self):
-        return self.sol[2].T[:, : self.rank]
-
-    @property
-    def U(self):
-        return self._U / np.sqrt(self.rw)[:, None]
+    def __init__(self, X, Y):
```

Comment:
Apologies since I know we discussed this before but I couldn't find it: is there a rule/standard for when to use uppercase `Y` instead of `y`?

Reply:
Again, this is a difference in terminology between `sklearn`  and `vegan` conventions.  Because `ConstrainedOrdination` is more or less a port of vegan, I felt like keeping the matrix variables upper-cased was important to follow those conventions.   I'm not as strident on this one, though, and could be convinced to change `Y` to `y`.

Reply:
I might suggest going lowercase for consistency if this was in the public API, but since it's just an implementation within the transformer, sticking with the `vegan` conventions is fine with me!